File: problems/0003-longest-substring-without-repeating-characters/typescript/solution.ts
// Longest Substring Without Repeating Characters — https://leetcode.com/problems/longest-substring-without-repeating-characters/
export function lengthOfLongestSubstring(s: string): number {
  const last = new Map<string, number>();
  let start = 0;
  let best = 0;
  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    const j = last.get(c);
    if (j !== undefined && j >= start) start = j + 1;
    last.set(c, i);
    best = Math.max(best, i - start + 1);
  }
  return best;
}

File: problems/0003-longest-substring-without-repeating-characters/go/solution.go
package solution

// Longest Substring Without Repeating Characters — https://leetcode.com/problems/longest-substring-without-repeating-characters/
func lengthOfLongestSubstring(s string) int {
	last := make(map[byte]int)
	start, best := 0, 0
	for i := 0; i < len(s); i++ {
		c := s[i]
		if j, ok := last[c]; ok && j >= start {
			start = j + 1
		}
		last[c] = i
		if i-start+1 > best {
			best = i - start + 1
		}
	}
	return best
}
